src/main/java/duke/tasks/TaskList.java

```diff
@@ -0,0 +1,239 @@
+        index--;
+        if (index < 0 || index > size) {
+            System.out.println(String.format("Indices have to be positive and less than %d.", size));
+            return;
+        }
+        Task task = this.records.get(index);
+        task.markDone();
+        dw.setDone(index);
+        System.out.println(Ui.format("Nice! I've marked this task as done:\n" + task.toString()));
+        return;
+    }
+
+    /**
+     * Marks a task in the list as undone. (1-indexed)
+     *
+     * @param index Index of task marked.
+     */
+    public void unMark(int index) {
+        index--;
+        if (index < 0 || index > size) {
+            System.out.println(String.format("Indices have to be positive and less than %d.", size));
+            return;
+        }
+        Task task = this.records.get(index);
+        task.markUndone();
+        dw.setUndone(index);
+        System.out.println(Ui.format("OK, I've marked this task as not done yet:\n\n" + task.toString()));
+        return;
+    }
+
+    /**
+     * Inserts a generic task to the TaskList.
+     *
+     * @param name Task name to be added.
+     */
+    public void insertTask(String name) {
+        Task t = new Task(name);
+        this.records.add(t);
+        dw.addToDb(t);
+        System.out.println(Ui.format("added: " + t));
+    }
+
+    /**
+     * Inserts a TaskToDo to the TaskList.
+     *
+     * @param name Task name to be added.
+     */
+    public void insertToDo(String name) {
+        TaskToDo t = new TaskToDo(name);
+        this.records.add(t);
+        dw.addToDb(t);
+        System.out.println(Ui.format("Got it. I've added this task:\n" + t.toString()));
+    }
+
+    /**
+     * Inserts a TaskToDo from the database to the TaskList, does not invoke addToDb() in avoid infinite loop.
+     *
+     * @param name Task name to be added.
+     * @param isInitial Boolean value of whether the program is population the TaskList.
+     */
+    public void insertToDo(String name, boolean isInitial) {
+        TaskToDo t = new TaskToDo(name);
+        this.records.add(t);
+    }
+
+    /**
+     * Inserts a TaskDeadline from the database to the TaskList.
+     *
+     * @param name Task name to be added.
+     * @param time Deadline time in string, according to the dateTimeRegex format.
+     */
+    public void insertDeadline(String name, String time) throws DateTimeFormatException {
+        System.out.println(time);
+        try {
+            if (!time.matches(dateTimeRegex)) { // dd/mm/yy tttt
+                throw new DateTimeFormatException();
+            }
+            LocalDateTime dateTime = LocalDateTime.parse(time, formatter);
+            TaskDeadline d = new TaskDeadline(time, dateTime);
+            this.records.add(d);
+            dw.addToDb(d);
+            System.out.println(Ui.format("Got it. I've added this task:\n" + d.toString()));
+        } catch (DateTimeFormatException e) {
+            System.out.println(e.getMessage());
+        }
```

Comment:
Would it be better to throw the Exception and catch it where the method is called?